json-schema-processor/src/main/java/io/micronaut/jsonschema/visitor/JsonSchemaVisitor.java

```diff
@@ -0,0 +1,220 @@
+ * limitations under the License.
+ */
+package io.micronaut.jsonschema.visitor;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.micronaut.core.annotation.AnnotationValue;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.inject.ast.ClassElement;
+import io.micronaut.inject.ast.EnumElement;
+import io.micronaut.inject.ast.PropertyElement;
+import io.micronaut.inject.ast.TypedElement;
+import io.micronaut.inject.visitor.TypeElementVisitor;
+import io.micronaut.inject.visitor.VisitorContext;
+import io.micronaut.inject.writer.GeneratedFile;
+import io.micronaut.jsonschema.JsonSchema;
+import io.micronaut.jsonschema.visitor.JsonSchemaConfigurationVisitor.JsonSchemaContext;
+import io.micronaut.jsonschema.visitor.aggregator.DocumentationInfoAggregator;
+import io.micronaut.jsonschema.visitor.aggregator.JacksonInfoAggregator;
+import io.micronaut.jsonschema.visitor.aggregator.SchemaInfoAggregator;
+import io.micronaut.jsonschema.visitor.aggregator.ValidationInfoAggregator;
+import io.micronaut.jsonschema.visitor.model.Schema;
+import io.micronaut.jsonschema.visitor.model.Schema.Type;
+import io.micronaut.jsonschema.visitor.serialization.JsonSchemaMapperFactory;
+
+import java.io.IOException;
+import java.io.Writer;
+import java.net.URI;
+import java.util.*;
+
+import static io.micronaut.jsonschema.visitor.JsonSchemaConfigurationVisitor.JSON_SCHEMA_CONFIGURATION_PROPERTY;
+
+/**
+ * A visitor for creating JSON schemas for beans.
+ * The bean must have a {@link JsonSchema} annotation.
+ *
+ * @since 1.0.0
+ * @author Andriy Dmytruk
+ */
+@Internal
+public final class JsonSchemaVisitor implements TypeElementVisitor<JsonSchema, Object> {
+
+    private static final List<SchemaInfoAggregator> SCHEMA_INFO_AGGREGATORS = List.of(
+        new JacksonInfoAggregator(),
+        new ValidationInfoAggregator(),
+        new DocumentationInfoAggregator()
+    );
+    private static final String SUFFIX = ".schema.json";
+    private static final String SLASH = "/";
+
+    @Override
+    public @NonNull TypeElementVisitor.VisitorKind getVisitorKind() {
+        return VisitorKind.AGGREGATING;
+    }
+
+    @Override
+    public void visitClass(ClassElement element, VisitorContext visitorContext) {
+        if (element.hasAnnotation(JsonSchema.class)) {
+            JsonSchemaContext context = visitorContext.get(JSON_SCHEMA_CONFIGURATION_PROPERTY, JsonSchemaContext.class, null);
+            if (context == null) {
+                context = JsonSchemaContext.createDefault();
+                visitorContext.put(JSON_SCHEMA_CONFIGURATION_PROPERTY, context);
+            }
+            createTopLevelSchema(element, visitorContext, context);
+        }
+    }
+
+    public static Schema createTopLevelSchema(TypedElement element, VisitorContext visitorContext, JsonSchemaContext context) {
+        Schema schema = context.createdSchemasByType().get(element.getGenericType().getName());
```

Comment:
it is likely that this creates dependencies between different classes, it should probably just check if the return value of `getGenericType` has the `@JsonSchema` annotation or not instead of sharing this state as this approach may not work if the order of compilation changes